File: bench/sort_window.py
#!/usr/bin/env python3
# benchmark-contract: sort-window;seed=123456789;iterations=2000000;width=8;algorithm=bubble
#
# sort_window — 2M iterations of "derive 8 values from the LCG state,
# bubble-sort them, fold the extremes back into the state". 8 passes of
# up to 7 compare-and-maybe-swap steps each: a compare/branch/store mill
# with a loop-carried dependency through the state.
#
# Contract: print exactly one line — the checksum, masked to 63 bits.
MASK = 0xFFFFFFFF

window = [0, 1, 2, 3, 4, 5, 6, 7]
state = 123456789

for i in range(2_000_000):
    state = (state * 1664525 + 1013904223) & MASK

    window[0] = state
    window[1] = state ^ 0xA5A5A5A5
    window[2] = (state + i) & MASK
    window[3] = state * 3
    window[4] = (state - i) & MASK
    window[5] = state >> 3
    window[6] = state << 1
    window[7] = (state + 7) & MASK

    # Textbook bubble sort over 8 slots, kept deliberately naive so
    # every language runs the same instruction mill.
    for p in range(8):
        for j in range(7 - p):
            if window[j] > window[j + 1]:
                window[j], window[j + 1] = window[j + 1], window[j]

    state = state ^ window[0] ^ window[7]

print(state & 0x7FFFFFFFFFFFFFFF)

File: bench/sort_window.js
// benchmark-contract: sort-window;seed=123456789;iterations=2000000;width=8;algorithm=bubble
//
// sort_window — 2M iterations of "derive 8 values from the LCG state,
// bubble-sort them, fold the extremes back into the state".
//
// Representation: Numbers. Two of the derived slots (`state * 3` and
// `state << 1`) are 64-bit values in the peers, but they stay under
// 2^36 here, which a double holds exactly — so only the XORs need help,
// since JS's `^` is defined on int32. `xor64` does the XOR on the two
// 32-bit halves and reassembles, which is exact for anything under 2^53.
// `state >> 3` and `state << 1` likewise become divide/multiply so they
// do not truncate to 32 bits the way JS's shift operators would.
//
// Contract: print exactly one line — the checksum, masked to 63 bits.
const ITERATIONS = 2000000;
const TWO32 = 4294967296;

function xor64(a, b) {
    const high = (Math.floor(a / TWO32) ^ Math.floor(b / TWO32)) >>> 0;
    const low = ((a >>> 0) ^ (b >>> 0)) >>> 0;
    return high * TWO32 + low;
}

const window8 = [0, 1, 2, 3, 4, 5, 6, 7];
let state = 123456789;

for (let i = 0; i < ITERATIONS; i++) {
    state = (Math.imul(state, 1664525) + 1013904223) >>> 0;

    window8[0] = state;
    window8[1] = xor64(state, 0xa5a5a5a5);
    window8[2] = (state + i) >>> 0;
    window8[3] = state * 3;
    window8[4] = (state - i) >>> 0;
    window8[5] = Math.floor(state / 8);
    window8[6] = state * 2;
    window8[7] = (state + 7) >>> 0;

    for (let pass = 0; pass < 8; pass++) {
        for (let j = 0; j + 1 < 8 - pass; j++) {
            if (window8[j] > window8[j + 1]) {
                const tmp = window8[j];
                window8[j] = window8[j + 1];
                window8[j + 1] = tmp;
            }
        }
    }

    state = xor64(xor64(state, window8[0]), window8[7]);
}

console.log(state.toString());
